New README and skeleton of documentation on docsy site.

ztreinhart opened the pull request:
This PR is a draft of an updated README and the first few pages of (hopefully) useful information on our documentation site (h/t @bradbeam). These pages include a getting started page and a reference for operation and command line flags.  I welcome all comments and constructive criticism. I also have a few specific questions:
* Do we want to provide a specific point of contact for VHS at StormForge? What will it be, if so? That could be added easily in this PR.
* What are people's thoughts about badges in the README? I got these from some README template, and I think they look nice, but some things I saw while researching GitHub READMEs indicate that people have surprisingly strong opinions about badges...
* Are we ready to roll out contributor guidelines and that sort of thing? Probably a question for @qedrakmar.

I will update the Architecture page with more details on a separate PR. Please suggest any additional pages we should have on our docs site, as well.

qedrakmar commented:
To your questions, @ztreinhart:

1) The points of contact should be here in GitHub, and through the Google Groups at https://groups.google.com/g/vhs-pre-rename-launch.  Acknowledging that this is a community project, and our contributions are as part of that community, not as a StormForge project, is important for the mindset of making sure this is properly released into the wild.  Once the community grows, we can add additional groups for maintainers, steering, and the like, if needed.

3) Initial Contributor Guidelines should be done this week, along with our Code of Conduct.  It is worth noting that we will not be using a CLA, and will instead be following the DCO model, so once that's in place, we will need ensure all PRs follow suit.

(I skipped 2, as I have no opinion on badges)

bradbeam commented on site/content/Architecture/_index.md:
```diff
@@ -0,0 +1,16 @@
+---
+
+## Introduction
+`vhs` is designed for flexibility and operates on the concept of a data flow that originates with a [source](#sources)
```
I think these links are broken?

ztreinhart replied:
Good catch. Fixed.

bradbeam commented on site/content/Getting Started/_index.md:
```diff
@@ -0,0 +1,139 @@
+    Taking VHS for a spin.
+---
+
+The quickest way to see `vhs` in action is to use the development Docker image in this repo to run a demo.
```
Since this is largely a copy of what's in the readme, I wonder if we should just have the readme link to this page instead

ztreinhart replied:
I think there might be some value in having a super simple demo for the README, but for now I've changed this to just link to the getting started page on the docs site.

bradbeam commented on site/content/Getting Started/_index.md:
```diff
@@ -0,0 +1,139 @@
+`./vhs --input "tcp|http" --address 0.0.0.0:80 --capture-response --prometheus-address 0.0.0.0:8888 --flow-duration 60m`
+
+The above command will capture live HTTP traffic on port 80 and calculate 
+[RED metrics](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/) for the captured
```
Probably worth listing out the metrics + comments here.

ztreinhart replied:
I've added a bit more of a description of the metrics provided as well as link to the reference section on metrics that describes all the metrics in much more detail.

bradbeam commented on site/content/Reference/_index.md:
```diff
@@ -0,0 +1,248 @@
+and responses from the TCP data, formats them as JSON, and prints them to the standard output.
+
+## Specifying Inputs and Outputs
+The core command line flags for `vhs` are focussed on defining the data flow that `vhs` will use for a recording/replay
```
sp focused

ztreinhart replied:
Good catch.

bradbeam commented on site/content/Reference/_index.md:
````diff
@@ -0,0 +1,248 @@
+```--input "<source|modifier(s)|format>"```
+
+Inputs are specified in a pipe-delimited (`|`), double-quoted string following the `--input` command line flag. 
+The first element in the string must specify a [source](#sources), followed by a pipe character, followed by zero or
+more [modifiers](#input-modifiers) separated by pipe characters, followed by another pipe character, and ending with an
+input [format](#input-formats) specifier. The specified source originates a data stream that is modified by the 
+specified modifiers and then formatted, or interpreted by the specified input format.
````
This feels overly verbose, followed by a pipe character. Followed by another pipe character. :)

Since we've already mentioned the input is pipe delimited, maybe something along the lines of `This must begin with an input [source](#sources), optionally contain [modifiers](#input-modifiers), and end with an [input format](#input-formats).`

ztreinhart replied:
`"Good|call."`

bradbeam commented on site/content/Reference/_index.md:
```diff
@@ -0,0 +1,248 @@
+In the example command given above, the input specifier is `--inputs "tcp|http"` where `tcp` specifies the TCP source
+and `http` specifies the HTTP input format. This example does not use any input modifiers. 
+
+The next sections will detail the currently available [input sources](#input-sources), [modifiers](#input-modifiers),
+and [formats](#input-formats) in `vhs`. Each source, modifier, and format may require additional configuration in the
+form of additional command line flags. These will be described where applicable.
```
May be worth dropping this.

ztreinhart replied:
👍

ztreinhart replied:
Dropped.

bradbeam commented on site/content/Reference/_index.md:
```diff
@@ -0,0 +1,248 @@
+
+##### `file`
+The `file` source reads data from a file on the local filesystem. It requires the following command line flag
+for configuration.
```
Probably worth mentioning what input filetypes are supported.

ztreinhart replied:
Hmm, could be useful!

ztreinhart replied:
Added a little more description of how the file source works and the supported file types.

bradbeam commented on site/content/Reference/_index.md:
```diff
@@ -0,0 +1,248 @@
+* `--gcs-bucket-name <GCS bucket name>` Required. Name of bucket that contains the object to be read.
+* `--gcs-object-name <object name>` Required. Name of object to be read.
+
+Note that this source also requires a JSON key file containing Google Cloud authentication credentials.
```
Might be worth mentioning the use of `GOOGLE_APPLICATION_CREDENTIALS` to specify the location of the credentials.

ztreinhart replied:
👍

ztreinhart replied:
Added.

bradbeam commented on site/content/Reference/_index.md:
```diff
@@ -0,0 +1,248 @@
+---
```
I wonder if it would be worth breaking this up into individual topic pages - input / output / modifiers / prometheus.

ztreinhart replied:
You might be right about that, but I think I'm going to leave that for a future revision.

bradbeam commented on site/content/Reference/_index.md:
```diff
@@ -0,0 +1,248 @@
+
+The provided metrics include measures of request rate, error rate, and request duration, sufficient for implementing 
+the [RED method](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/) of 
+microservice monitoring.  Metrics are supplied on a [Prometheus](https://prometheus.io/) endpoint. Request counts are
+available in a [counter vector](https://prometheus.io/docs/concepts/metric_types/#counter) labeled with HTTP method,
+status code, and path. Request durations are available in a 
+[summary vector](https://prometheus.io/docs/concepts/metric_types/#summary) with percentiles at 50% +/- 5%, 75% +/- 1%,
+90% +/- 0.5%, 95% +/- 0.5%, 99% +/- 0.1%, 99.9% +/- 0.01%, and 99.99% +/- 0.001%. Durations are also labeled with HTTP
+method, status code, and path.
```
A quick table breakdown might be best

ztreinhart replied:
Yep, that's a worthwhile improvement.

ztreinhart replied:
Added in most recent commit.

bradbeam commented on site/content/Getting Started/_index.md:
````diff
@@ -0,0 +1,139 @@
+Once you have Docker set up and working, you can build the development container (which installs some helpful tools
+and mounts the source in the container) by changing directory into the repository you cloned and running:
+
+```
````
Might be worth adding a note or two about what `make dev` is doing.

ztreinhart replied:
Added a description of the make dev process and the development container.

ztreinhart commented on README.md:
```diff
@@ -1,32 +1,116 @@
+
+
+
+<!-- ABOUT THE PROJECT -->
```
Done.

ztreinhart commented on README.md:
```diff
@@ -1,32 +1,116 @@
+
+`vhs` is a versatile tool for network traffic capture. `vhs` can be run as a command line tool or be deployed into your
+Kubernetes cluster as a sidecar, where it can capture traffic to and from your services. Captured traffic
+can be analyzed to produce live Prometheus metrics or saved for use in offline analysis, load testing, or whatever
```
Good call, I'll clarify this part.

ztreinhart commented on README.md:
```diff
@@ -1,32 +1,116 @@
 
-Open two bash sessions on the container. 
+### Demo Setup
+Open a bash session on the container by running the following command in a terminal:
```
Yeah, might as well head off any :trollface:-ing where possible.

ztreinhart commented on README.md:
````diff
@@ -1,32 +1,116 @@
+In this new session, run the following command to build `vhs` and run it to capture our local HTTP traffic.
 
 ```
 $ go build ./cmd/vhs && ./vhs --input "tcp|http" --output "json|stdout" --capture-response --address 0.0.0.0:1111 --middleware ./testdata/http_middleware.bash | jq -R "fromjson | .body" 2> /dev/null
````
Good idea, I think we'll handle this in a separate PR and update the docs when we've cleaned up our Dockerfile situation.

ztreinhart commented on README.md:
```diff
@@ -41,23 +125,80 @@ Sample output:
+<!-- CONTACT -->
+## Contact
+
+Your Name - [@twitter_handle](https://twitter.com/twitter_handle) - email
```
Since Noah has gotten a mailing list and public Slack set up, I'll link to those here.

ztreinhart commented on site/content/Architecture/_index.md:
```diff
@@ -0,0 +1,16 @@
+---
+title: "Architecture"
+linkTitle: "Architecture"
```
Done.

donovanmuller commented on README.md:
````diff
@@ -1,32 +1,116 @@
 
-In one, run this script to boot a simple echo server and a curl request that calls the server every second:
+```
+$ docker exect -it vhs_dev bash
````
```suggestion
$ docker exec -it vhs_dev bash
```

ztreinhart replied:
Based on other comments above, the demo section of the README will now just link to the getting started page on the docs site, so this section will go away.

donovanmuller commented on README.md:
````diff
@@ -1,32 +1,116 @@
+Open another bash session on the container in a new terminal:
+
+```
+$ docker exect -it vhs_dev bash
````
```suggestion
$ docker exec -it vhs_dev bash
```

ztreinhart replied:
Replaced by a link to the getting started section on the docs site.

donovanmuller commented on site/content/Getting Started/_index.md:
````diff
@@ -0,0 +1,139 @@
+Open a bash session on the container by running the following command in a terminal:
+
+```
+$ docker exect -it vhs_dev bash
````
```suggestion
$ docker exec -it vhs_dev bash
```

ztreinhart replied:
Good catch, fixed! Thanks!

donovanmuller commented on site/content/Getting Started/_index.md:
````diff
@@ -0,0 +1,139 @@
+Open another bash session on the container in a new terminal:
+
+```
+$ docker exect -it vhs_dev bash
````
```suggestion
$ docker exec -it vhs_dev bash
```

ztreinhart replied:
Fixed.

zeisss commented on README.md:
````diff
@@ -1,32 +1,116 @@
+In this new session, run the following command to build `vhs` and run it to capture our local HTTP traffic.
 
 ```
 $ go build ./cmd/vhs && ./vhs --input "tcp|http" --output "json|stdout" --capture-response --address 0.0.0.0:1111 --middleware ./testdata/http_middleware.bash | jq -R "fromjson | .body" 2> /dev/null
````
fyI: prefixing commands with `$` normally (according to markdownlinter) means you show the full output of the command. If you just want to show the command as is, without any output.

See https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md014---dollar-signs-used-before-commands-without-showing-output

ztreinhart replied:
Fixed in most recent commit.

ztreinhart commented:
Okay, I believe [815600c](https://github.com/rename-this/vhs/commit/815600c04b32290564cdada2d09850ffce057ced) addresses the comments above, except for the ones I specifically suggested we handle in another PR.
